executor/analyze.go

```diff
@@ -198,14 +154,77 @@ func (e *AnalyzeExec) Next(ctx context.Context, _ *chunk.Chunk) error {
 	if err != nil {
 		return err
 	}
+
+	// Update analyze options to mysql.analyze_options for auto analyze.
 	err = e.saveV2AnalyzeOpts()
 	if err != nil {
-		e.Ctx().GetSessionVars().StmtCtx.AppendWarning(err)
+		sessionVars.StmtCtx.AppendWarning(err)
+	}
+
+	if sessionVars.InRestrictedSQL {
+		return statsHandle.Update(infoSchema)
 	}
-	if e.Ctx().GetSessionVars().InRestrictedSQL {
-		return statsHandle.Update(e.Ctx().GetInfoSchema().(infoschema.InfoSchema))
+
+	return statsHandle.Update(infoSchema, cache.WithTableStatsByQuery())
+}
+
+// filterAndCollectTasks filters the tasks that are not locked and collects the table IDs.
+func filterAndCollectTasks(tasks []*analyzeTask, statsHandle *handle.Handle, infoSchema infoschema.InfoSchema) ([]*analyzeTask, map[int64]struct{}, []string) {
+	var filteredTasks []*analyzeTask
+	tids := make(map[int64]struct{})
+	var skippedTables []string
+
+	for _, task := range tasks {
+		tableID := getTableIDFromTask(task)
+		if !statsHandle.IsTableLocked(tableID) {
+			filteredTasks = append(filteredTasks, task)
+		}
+		if _, ok := tids[tableID]; !ok {
+			if statsHandle.IsTableLocked(tableID) {
+				tbl, ok := infoSchema.TableByID(tableID)
+				if !ok {
+					continue
```

Comment:
Why continue here, it's `return nil` before

Reply:
Good catch. This was a mistake.

Reply:
It's almost impossible to meet such cases in normal scenarios. Even if the table is dropped during the execution of this statement, we should still be able to get the table from the `infoschema` here.

I agree that the previous `return nil` is wrong.
Personally, I think DEBUG level logging is basically useless since the default is INFO level, and I would suggest that we directly record it like `table id: 123` to the `skippedTables`.

Reply:
The reason I didn't generate a statement warning here is that we can't get the table, so if we use TableID here then it would be too confusing for users.

Reply:
> It's almost impossible to meet such cases in normal scenarios. Even if the table is dropped during the execution of this statement, we should still be able to get the table from the `infoschema` here.

So let's change it to a warning. That would be helpful.

Reply:
LGTM.